Fixed #28211 -- Stopped combining Q objects if either side is falsey.

orf opened the pull request:
Perhaps a good workaround for [#28211](https://code.djangoproject.com/ticket/28211) is to just not OR/AND Q objects unless each side is truthy.

`Q() | Q(x=1)` can be reduced down to `Q(x=1)`, and `Q() & Q(x=1)` can be reduced to just `Q(x=1)` (as far as I can tell?), so we could just reduce both these cases down to `Q(x=1)` in the `Q._combine()` function.

charettes commented:
That's a good idea! My only concern would be that some code might indirectly expect `Q` combination to return a new instance. I suggest we return copies of `self` or `other` instead to be extra safe.

MatthewWilkes commented on tests/queries/test_q.py:
```diff
@@ -3,6 +3,30 @@
+    def test_and_falsey(self):
+        truthy = Q(x=1)
+        combined = truthy & Q()
+        self.assertTrue(combined is truthy)
```
I believe you'll want to use `self.assertIs` so there are better error messages, and for consistency with other parts of the test suite.

orf replied:
Good catch, thanks!

orf commented:
Thanks @charettes! I'm sort of against that idea (I can't imagine a situation it would happen, and it's slower 👎 ). But I can see it's good to keep the implicit contract so I've done it by deepcopying the arguments before returning.

charettes commented:
@orf yeah I'm not a fan of it myself but I think it's safer this way.

felixxm commented on django/db/models/query_utils.py:
```diff
@@ -61,6 +62,14 @@ def __init__(self, *args, **kwargs):
+            return copy.deepcopy(other)
+        elif not other:
+            return copy.deepcopy(self)
+
```
I think that we can squash this to two cases, i.e.:
```python
if not self and other:
    return copy.deepcopy(other)
elif not other:
    return copy.deepcopy(self)
```

Naddiseo replied:
or more simply:

```python
if not other:
    return copy.deepcopy(self)
elif not self: # bool(other) is implied to be True by the previous condition
    return copy.deepcopy(other)

orf commented:
Ok, I've squashed the conditionals down and added a comment in case anyone in the future is confused.

Should I add something to the release notes, or does this count as an internal change that's not documented?

charettes commented:
No release notes should be needed as this is a bug fix that won't be backported.

timgraham commented on tests/queries/test_q.py:
```diff
@@ -3,6 +3,30 @@
 
 
 class QTests(SimpleTestCase):
+    def test_and_falsey(self):
```
While making a few cosmetic edits, I noticed that this test is passing without the fix. If it's just adding additional test coverage, I'll merge it as a separate commit.

orf replied:
Hmm, I feel that it shouldn't pass. An unbalanced OR with one side being `x` shouldn't equal just `x`. IMO the first case shouldn't have been possible in the first place, but it is an independent and different thing than just `x`.

timgraham replied:
Looks like it's passing because of https://github.com/django/django/blob/b3eb6eaf1a197ff155faf333871da032c77ba855/django/utils/tree.py#L95-L104.
